src/btreemap.rs

```diff
@@ -930,6 +930,56 @@ where
         }
     }
 
+    /// Returns an interator pointing to the first element below the given bound.
+    /// Returns an empty iterator if there are no keys below the given bound.
+    pub fn iter_upper_bound(&self, bound: &K) -> Iter<K, V, M> {
+        if self.root_addr == NULL {
+            // Map is empty.
+            return Iter::null(self);
+        }
+
+        let mut cursors = vec![];
+        let mut node = self.load_node(self.root_addr);
+        loop {
+            match node.keys.binary_search(bound) {
+                Ok(idx) | Err(idx) => {
+                    match node.node_type {
+                        NodeType::Leaf => {
+                            if idx > 0 {
+                                cursors.push(Cursor::Node {
+                                    node,
+                                    next: Index::Entry(idx - 1),
+                                });
+                                // Leaf node. Return an iterator with the found cursors.
+                                return Iter::new_in_range(
+                                    self,
+                                    (Bound::Unbounded, Bound::Unbounded),
+                                    cursors,
+                                );
+                            } else {
+                                // The upper bound is less than or equal to the first key in the map.
+                                // We return an empty iterator.
```

Comment:
This logic is too simplistic. The problem with upper bounds is that we might have to backtrack.
If we see an internal node with keys [1, 10] and look for upper bound 5, we might descend into the slot between 1 and 10 and find keys [6, 8, 9] there, meaning that we should have taken 1.

That never happens with `range`, so we need a slightly different approach for the upper bound.

Reply:
So you mean there's a bug really here? We shouldn't return null but rather keep looping through so that we catch 1 in your example?

Reply:
Yes, there was a bug. I wrote a more thorough test (checking upper bounds for _all_ keys, not just the last key) and found the bug. It's fixed now!